aperturedb/Connector.py

```diff
@@ -152,39 +160,44 @@ def _authenticate(self, user, password="", token=""):
         if session_info["status"] != 0:
             raise Exception(session_info["info"])
 
-        self.session = Session(session_info["session_token"],
-                               session_info["refresh_token"],
-                               session_info["session_token_expires_in"],
-                               session_info["refresh_token_expires_in"],
-                               )
+        self.shared_data.session = Session(session_info["session_token"],
+                                           session_info["refresh_token"],
+                                           session_info["session_token_expires_in"],
+                                           session_info["refresh_token_expires_in"],
+                                           time.time()
+                                           )
 
     def _check_session_status(self):
-
-        if not self.session:
+        if not self.shared_data.session:
             return
 
-        if not self.session.valid():
+        if not self.shared_data.session.valid():
```

Comment:
do we want to think about locking around the test? It could be possible to have a thread start refreshing, then have another enter and do another refresh.

Or perhaps we should retest after locking in `_refresh_token`?

Reply:
The _refresh_token is only called from within _check_session_status, via _renew_session, which has the logic to try this in case of any failures.
I chose to implement it this way so that we do not have a lock around a bigger section.